File: examples/python/basic-langchain-agent/src/callback.py
from __future__ import annotations
from typing import Any

from langchain_core.messages import AIMessage, ToolMessage
from langchain_core.callbacks.base import BaseCallbackHandler
from langchain_core.outputs.llm_result import LLMResult
from uuid import UUID


class CallbackHandler(BaseCallbackHandler):
    """A simple callback handler for langchain invoke callbacks that just prints out the events."""

    def on_tool_start(
        self,
        serialized: dict[str, Any],
        input_str: str,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        inputs: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> Any:
        print(f"[Callback] Tool started: {kwargs['tool_call_id']}\n")

    def on_tool_end(
        self,
        output: ToolMessage,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        **kwargs: Any,
    ) -> Any:
        print(f"[Callback] Tool ended: {output.tool_call_id}\n")

    def on_llm_start(
        self,
        serialized: dict[str, Any],
        prompts: list[str],
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        tags: list[str] | None = None,
        metadata: dict[str, Any] | None = None,
        **kwargs: Any,
    ) -> Any:
        print("[Callback] LLM started\n")

    async def on_llm_end(
        self,
        response: LLMResult,
        *,
        run_id: UUID,
        parent_run_id: UUID | None = None,
        tags: list[str] | None = None,
        **kwargs: Any,
    ) -> None:
        print(f"[Callback] LLM ended - {print_tool_calls_from_llm_result(response)}\n")


# function that takes in LLMResult, iterates through generations and prints tool call name and tool call id pairings.
def print_tool_calls_from_llm_result(llm_result: LLMResult) -> list[str]:
    """return dictionary of tool call name and tool call id pairings from LLMResult."""
    # result: dict[str,str] = {}
    result: list[str] = []
    for generation_list in llm_result.generations:
        for generation in generation_list:
            # check if generation is ChatGeneration to access message attribute
            if not hasattr(generation, "message"):
                continue
            message = generation.message  # type: ignore

            # check if message is BaseMessage to access tool_calls attribute
            if not isinstance(message, AIMessage):
                continue

            tool_calls = message.tool_calls
            for tool_call in tool_calls:
                tool_name = tool_call.get("name")
                tool_id = tool_call.get("id")
                result.append(f"{tool_name}:{tool_id}")
    return result

File: examples/typescript/basic-langchain-agent/src/callback.ts
import { BaseCallbackHandler } from '@langchain/core/callbacks/base'
import type { Serialized } from '@langchain/core/load/serializable'
import { AIMessage } from '@langchain/core/messages'
import type { Generation, LLMResult } from '@langchain/core/outputs'

export class CallbackHandler extends BaseCallbackHandler {
  name = 'CallbackHandler'

  async handleToolStart(
    _tool: Serialized,
    _input: string,
    _runId: string,
    _parentRunId?: string,
    _tags?: string[],
    metadata?: Record<string, unknown>,
    name?: string,
  ): Promise<void> {
    console.log(`[Callback] Tool started: ${metadata?.tool_call_id || name}\n`)
  }

  async handleToolEnd(
    _output: string,
    runId: string,
    _parentRunId?: string,
    _tags?: string[],
  ): Promise<void> {
    console.log(`[Callback] Tool ended: ${runId}\n`)
  }

  async handleLLMStart(
    _llm: Serialized,
    _prompts: string[],
    _runId: string,
    _parentRunId?: string,
    _extraParams?: Record<string, unknown>,
    _tags?: string[],
    _metadata?: Record<string, unknown>,
  ): Promise<void> {
    console.log('[Callback] LLM started\n')
  }

  async handleLLMEnd(
    output: LLMResult,
    _runId: string,
    _parentRunId?: string,
    _tags?: string[],
  ): Promise<void> {
    const toolCalls = printToolCallsFromLLMResult(output)
    console.log(`[Callback] LLM ended - ${toolCalls.join(', ')}\n`)
  }
}

function printToolCallsFromLLMResult(llmResult: LLMResult): string[] {
  const result: string[] = []

  for (const generation of llmResult.generations) {
    for (const gen of generation) {
      const generationWithMessage = gen as Generation & { message?: AIMessage }
      if (
        generationWithMessage.message &&
        generationWithMessage.message instanceof AIMessage
      ) {
        const message = generationWithMessage.message
        if (message.tool_calls) {
          for (const toolCall of message.tool_calls) {
            result.push(`${toolCall.name}:${toolCall.id}`)
          }
        }
      }
    }
  }

  return result
}
